### ISSUE TITLE
Dataloader is broken. It doesn't work when shuffle is True.

### ISSUE THREAD

**Issue by Abdul-Mukit** [WongKinYiu/YOLO#85]
## Describe the bug
When shuffle is set to True dataloader generates random target data.

## How to reproduce.
Try the codes and use the following commands to see output.
Code:
```
def main(cfg: Config):
    progress = ProgressLogger(cfg, exp_name=cfg.name)
    device, use_ddp = get_device(cfg.device)
    dataloader = create_dataloader(cfg.task.data, cfg.dataset, cfg.task.task, use_ddp)
    print(f"Dataloader created with total samples: {len(dataloader.dataset)}.")
    print(f"Dataloader.batchsize={dataloader.batch_size}")

    print("Iterating dataloader")
    batch_size, batch_images, batch_targets, batch_reverse, batch_path = next(iter(dataloader))
    print(f'batch_path={batch_path}')
    print("batch_targets[0]:")
    print(batch_targets[0])
    print(batch_targets[0].shape)

    
    print('Directly reading from dataloader.dataset.data')
    for item_path, item_targets  in dataloader.dataset.data:
        if not item_path == batch_path[0]:
            continue
        print(f'item_path={item_path}')
        print("item_targets:")
        print(item_targets)
        print(item_targets.shape)
```

Try the command:
`python yolo/dataloader_debug.py task=train model=v9-s dataset=mock task.epoch=1 device=cpu cpu_num=0 task.data.batch_size=1 task.data.shuffle=True lucky_number=4`

```
Dataloader created with total samples: 5.
Dataloader.batchsize=1
Iterating dataloader
batch_path=(PosixPath('tests/data/images/train/000000050725.jpg'),)
batch_targets[0]:
tensor([[ 42.0000,   0.0000,   0.0000,   5.1500, 122.9200],
        [ 51.0000,   0.0000, 116.5900,   4.5800, 169.5700],
        [ 56.0000, 136.4000,  83.9200, 221.1400, 257.1200],
        [ 56.0000, 395.3500,  81.5700, 498.7400, 251.4100],
        [ 60.0000,  20.0000, 237.9300, 640.0000, 389.0300],
        [ 58.0000, 200.1700, 112.2100, 407.4700, 321.4400],
        [ 41.0000, 484.7200, 223.6600, 555.0200, 309.7300],
        [ 41.0000, 418.2800, 219.0700, 486.5900, 303.8200],
        [ 75.0000, 248.3900, 230.0500, 371.6900, 326.1200],
        [ 56.0000, 205.4700,  74.9100, 284.0100, 165.6500],
        [ 56.0000, 424.8700,  76.3700, 461.5800, 167.3400],
        [ 56.0000, 136.4000, 516.9200, 221.1400, 640.0000],
        [ 56.0000, 395.3500, 514.5700, 498.7400, 640.0000],
        [ 58.0000, 200.1700, 545.2100, 407.4700, 640.0000],
        [ 56.0000, 205.4700, 507.9100, 284.0100, 598.6500],
        [ 56.0000, 424.8700, 509.3700, 461.5800, 600.3400]])
torch.Size([16, 5])
Directly reading from dataloader.dataset.data
item_path=tests/data/images/train/000000050725.jpg
item_targets:
tensor([[42.0000,  0.7554,  0.0000,  0.9691,  0.5764],
        [50.0000,  0.6583,  0.6437,  0.8583,  0.7868],
        [50.0000,  0.7310,  0.5784,  0.8972,  0.6937],
        [47.0000,  0.0569,  0.4459,  0.2300,  0.6665],
        [50.0000,  0.5461,  0.4780,  0.7870,  0.6194],
        [51.0000,  0.5670,  0.6495,  0.7552,  0.8292],
        [51.0000,  0.7450,  0.6887,  0.7823,  0.7579],
        [51.0000,  0.8707,  0.5665,  0.9679,  0.6493],
        [51.0000,  0.8052,  0.5414,  0.8463,  0.5723],
        [51.0000,  0.6904,  0.6412,  0.7312,  0.6633],
        [51.0000,  0.7505,  0.7825,  0.7851,  0.8052]], dtype=torch.float64)
torch.Size([11, 5])
```
Dataloader seems to be completely broken. Where as the image 000000050725.jpg had 11 annotations, dataloader produced 16 annotations. Often the situation is reverse. For example, if the actual annotation count is 21, dataloader would produce 2 annotations with often both being -1 for class values. Meaning they are empty.

I have also observed that calling `dataloader.dataset[<any int number>]` produces different results at different times.

## Expected behavior
`python yolo/dataloader_debug.py task=train model=v9-s dataset=mock task.epoch=1 device=cpu cpu_num=0 task.data.batch_size=1 task.data.shuffle=False`
produced the correct output below. Notice the `shuffle=Falase`. 
If `shuffle=True`, the output should be like the following too.
```
Dataloader created with total samples: 5.
Dataloader.batchsize=1
Iterating dataloader
batch_path=[PosixPath('tests/data/images/train/000000050725.jpg')]
batch_targets[0]:
tensor([[ 42.0000, 448.5900,   0.0000, 551.1500, 306.9200],
        [ 50.0000, 401.9600, 349.9500, 497.9800, 441.5300],
        [ 50.0000, 436.8700, 308.2000, 516.6800, 381.9800],
        [ 47.0000, 113.3200, 223.3900, 196.3900, 364.5700],
        [ 50.0000, 348.1300, 243.9500, 463.7800, 334.4300],
        [ 51.0000, 358.1600, 353.6900, 448.5000, 468.7000],
        [ 51.0000, 443.6200, 378.7900, 461.4800, 423.0600],
        [ 51.0000, 503.9500, 300.5900, 550.5800, 353.5700],
        [ 51.0000, 472.4800, 284.5100, 492.2200, 304.2400],
        [ 51.0000, 417.3900, 348.3700, 437.0000, 362.5000],
        [ 51.0000, 446.2200, 438.8300, 462.8600, 453.3000]])
torch.Size([11, 5])
Directly reading from dataloader.dataset.data
item_path=tests/data/images/train/000000050725.jpg
item_targets:
tensor([[42.0000,  0.7554,  0.0000,  0.9691,  0.5764],
        [50.0000,  0.6583,  0.6437,  0.8583,  0.7868],
        [50.0000,  0.7310,  0.5784,  0.8972,  0.6937],
        [47.0000,  0.0569,  0.4459,  0.2300,  0.6665],
        [50.0000,  0.5461,  0.4780,  0.7870,  0.6194],
        [51.0000,  0.5670,  0.6495,  0.7552,  0.8292],
        [51.0000,  0.7450,  0.6887,  0.7823,  0.7579],
        [51.0000,  0.8707,  0.5665,  0.9679,  0.6493],
        [51.0000,  0.8052,  0.5414,  0.8463,  0.5723],
        [51.0000,  0.6904,  0.6412,  0.7312,  0.6633],
        [51.0000,  0.7505,  0.7825,  0.7851,  0.8052]], dtype=torch.float64)
torch.Size([11, 5])
```

## System Info (please complete the following ## information):
 - OS: [Debian]
 - Python Version: [3.9]
 - PyTorch Version: [2.4]
 - CUDA/cuDNN/MPS Version: [CUDA 12.1]
 - YOLO Model Version: [YOLOv9-s]

## Request to the Owners
Please prioritize performing sanity checks on a very simple training loop.
Use the same 5 images for training and validation. Train from scratch and if the model can't overfit in the simplest possible case, then there must be something very wrong. If the problem I presented above is really a problem, then the model will never converge. I think reordering priorities from DDP or other issues to first making sure the basic training pipeline works will be more helpful to the community. Then more people can contribute to accelerate the growth of this repo.

cc: @WongKinYiu @henrytsui000

**Comment by henrytsui000** [comment]
Hi,

Have you enabled [Mosaic](https://arxiv.org/pdf/2004.10934) and [RandomCrop](https://www.alamy.com/stock-photo-wet-and-dirty-rice-field-in-bali-indonesia-170946270.html)? According to the YOLOv4 paper, Mosaic augmentation combines 3 additional images into the current image, often leading to more than 11 bounding boxes. RandomCrop, on the other hand, randomly crops a region and treats it as ground truth.

I tried the following code:
```python
from yolo.tools.drawer import draw_bboxes
img = draw_bboxes(batch_images, batch_targets)
img.save('ImageCheck.png')
```
The box mismatch you're seeing is indeed caused by the combination of Mosaic and RandomCrop—it’s not a bug.

Best regards,  
Henry Tsui

**Comment by Abdul-Mukit** [comment]
@henrytsui000 Thank you. I got it now and will walk through the augmentation process myself.
I encountered a situation where I had batch size of 1, and due to the augmentations the target was empty or of size (1,0,5) = [batch_size, numb_targers, 5]. That caused this line to fail:
https://github.com/WongKinYiu/YOLO/blob/8228669808a626fc5f9c233fdb35550b5e041fae/yolo/utils/bounding_box_utils.py#L221

Can you please suggest what should be the solution?
Should I just return loss_each to all zeros in the `train_one_batch` just before this line by checking `if targets.shape[1] == 0`
https://github.com/WongKinYiu/YOLO/blob/8228669808a626fc5f9c233fdb35550b5e041fae/yolo/tools/solver.py#L70

Or should I return `loss_each` to all zeros from `DualLoss.__call__()`
https://github.com/WongKinYiu/YOLO/blob/8228669808a626fc5f9c233fdb35550b5e041fae/yolo/tools/loss_functions.py#L123
Or the solution should be completely different?
Thank you again for your time.

**Comment by Abdul-Mukit** [comment]
@henrytsui000 Thank you. I got it now and will walk through the augmentation process myself then close the issue.
I'll try creating a notebook example on augmentations. I appreciate your time.
